qiskit_machine_learning/datasets/ad_hoc.py

```diff
@@ -146,7 +146,11 @@ def ad_hoc_data(
         x = np.array(x)
         phi = np.sum(x[:, None, None] * z_i, axis=0)
         phi += sum(((np.pi - x[i1]) * (np.pi - x[i2]) * z_i[i1] @ z_i[i2] for i1, i2 in ind_pairs))
-        u_u = scipy.linalg.expm(1j * phi)  # pylint: disable=no-member
+        # u_u was actually scipy.linalg.expm(1j * phi), but this method is
+        # faster because phi is always a diagonal matrix.
+        # We first extract the diagonal elements, then exponentiate them, then
+        # construct a diagonal matrix from them.
+        u_u = np.diag(np.exp(1j * np.diag(phi)))
```

Comment:
`phi` is always a diagonal matrix, because it is constructed out of multiplication of indices of https://github.com/Qiskit/qiskit-machine-learning/blob/0c6055811abbf39e9e02771ac891616fc61a6d7b/qiskit_machine_learning/datasets/ad_hoc.py#L118 (i.e. `z_i`). We see in L148 that the off diagonals are always 0.